src/shims/x86/sha.rs

```diff
@@ -0,0 +1,209 @@
+                assert_eq!(b_len, 4);
+                assert_eq!(k_len, 4);
+                assert_eq!(dest_len, 4);
+
+                let mut a = [0; 4];
+                for (i, dst) in a.iter_mut().enumerate() {
+                    *dst = this.read_scalar(&this.project_index(&a_reg, i as u64)?)?.to_u32()?
+                }
+                let mut b = [0; 4];
+                for (i, dst) in b.iter_mut().enumerate() {
+                    *dst = this.read_scalar(&this.project_index(&b_reg, i as u64)?)?.to_u32()?
+                }
+                let mut k = [0; 4];
+                for (i, dst) in k.iter_mut().enumerate() {
+                    *dst = this.read_scalar(&this.project_index(&k_reg, i as u64)?)?.to_u32()?
+                }
+
+                let result = sha256rnds2_epu32(a, b, k);
+                for (i, part) in result.into_iter().enumerate() {
+                    this.write_scalar(Scalar::from_u32(part), &this.project_index(&dest, i as u64)?)?;
+                }
+            }
+            // Used to implement the _mm_sha256msg1_epu32 function.
+            "256msg1" => {
+                let [a, b] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+
+                let (a_reg, a_len) = this.operand_to_simd(a)?;
+                let (b_reg, b_len) = this.operand_to_simd(b)?;
+                let (dest, dest_len) = this.mplace_to_simd(dest)?;
+
+                assert_eq!(a_len, 4);
+                assert_eq!(b_len, 4);
+                assert_eq!(dest_len, 4);
+
+                let mut a = [0; 4];
+                for (i, dst) in a.iter_mut().enumerate() {
+                    *dst = this.read_scalar(&this.project_index(&a_reg, i as u64)?)?.to_u32()?
+                }
+                // least significant part of b
+                let b_ls = this.read_scalar(&this.project_index(&b_reg, 0)?)?.to_u32()?;
+
+                let result = sha256msg1_epu32(a, b_ls);
+                for (i, part) in result.into_iter().enumerate() {
+                    this.write_scalar(Scalar::from_u32(part), &this.project_index(&dest, i as u64)?)?;
+                }
+            }
+            // Used to implement the _mm_sha256msg2_epu32 function.
+            "256msg2" => {
+                let [a, b] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+
+                let (a_reg, a_len) = this.operand_to_simd(a)?;
+                let (b_reg, b_len) = this.operand_to_simd(b)?;
+                let (dest, dest_len) = this.mplace_to_simd(dest)?;
+
+                assert_eq!(a_len, 4);
+                assert_eq!(b_len, 4);
+                assert_eq!(dest_len, 4);
+
+                let mut a = [0; 4];
+                for (i, dst) in a.iter_mut().enumerate() {
+                    *dst = this.read_scalar(&this.project_index(&a_reg, i as u64)?)?.to_u32()?
+                }
+                // most significant part of b
```

Comment:
```suggestion
                // b_ms becomes the most significant part of b, i.e., b[2..4].
```
I first thought this would be the most significant half of every element or so. It's just the second half of the array, quite confusing to call that "most significant".

Reply:
Yeah, the whole thing is quite confusing because the registers are little endian. I guess I should mention it in the code.